### DISCUSSION TITLE
Will Mojo eventually include an optional JiT capability in mojo AoT programs?

### DISCUSSION THREAD

**Discussion by gdonval** [modular/modular#1157] [Ideas]
There are quite a few domains where being able to JIT compile computational kernels would make a lot of sense performance-wise either for capability expansion or in highly-dynamic-yet-specialized systems. 

For instance, if one wants to provide their users with a way to write a custom kernel and feed it as-is (i.e. in source form) to the main AoTed program.

Science and engineering are also full of cases where you basically want to "`autotune`" the very hot stuff at runtime right after a few important variables are set right at launch time. Telling people "please recompile after setting `alias my_important_var = N` to the value that applies to your case" is not great.

**Comment by stellarpower** [comment]
I was writing a tool in C++ to build a computational graph from a GUI and to JIT a kernel for executing that graph. Given the lack of introspection (yet) in the standard library, this has been a good part of the problem, since I want to avoid by-hand boilerplate in the tool I have been using up until now. And now I'm considering using Mojo for this, but it'd be essential in order to replicate that functionality that I could recompile the function implementing the graph based on what is in the UI, at runtime. I guess worst-case I could call out to the compiler as a process, embed mojo inside my application, etc., but it's an important usecase I feel.
